docs/decision/structs-interfaces.md

```diff
@@ -0,0 +1,231 @@
+}
+
+func myResourceCreate(ctx context.Context, req ResourceCreateRequest, res ResourceCreateResponse) {}
+```
+
+> The second form is conventional when constructing the `Resource` struct from the existing `helper/schema` framework.
+
+The variety of ways struct methods can be given a value makes the task of static code analysis harder, as it is harder to identify which fragment of code contains the body of the function belonging to `r.Create`. The convention of declaring named functions introduces many long function names to the global namespace. This is a **disadvantage**.
+
+
+#### Interfaces
+
+Receiver methods fulfilling the method set of an interface are always named the same: for example, in solution 2, an implementation of the `Resource` interface will define a method called `Create`. This is an **advantage** in the case of static analysis, since method names signify intent.
+
+### Documentation
+
+#### Structs
+
+When navigating documentation, struct value functions are not grouped at all. This is a **disadvantage**.
+
+> For example, see the documentation for [`helper/schema#CreateFunc`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/helper/schema#CreateFunc).
+
+#### Interfaces
+
+When navigating documentation, the functions of an interface are grouped together by their namespace. This is an **advantage**.
+
+> For example, see the documentation for [`net/http#Server`](https://pkg.go.dev/net/http#Server).
+
+### Compatibility
+
+#### Structs
+
+There is no check at compile time that a struct has non-nil values for every field.
+
+For example, given the following struct:
+
+```go
+type Resource struct {
+  Create func() error
+}
+```
+
+And an instance `r` of that struct type:
+
+```go
+r := Resource{}
+```
+
+A nil pointer dereference panic will occur at runtime if the SDK calls `r.Create()`. The SDK must check at runtime that struct functions are set appropriately, so developers do not get feedback at compile time if they have omitted a required function. This is a **disadvantage**.
+
+On the other hand, framework developers can add additional methods to an existing struct and document the old ones as deprecated, without forcing provider developers to use a specific combination of old and new methods. 
+
+> This was the approach taken in `helper/schema` when two new sets of CRUD functions were added: `CreateContext` etc ([#276](https://github.com/hashicorp/terraform-plugin-sdk/pull/276)), and `CreateWithoutTimeout` etc ([#723](https://github.com/hashicorp/terraform-plugin-sdk/pull/723)). 
+ 
+As is evident from the `helper/schema` case, such new methods must have unique and therefore rather verbose names within the existing `Resource` struct, putting the struct approach at a **slight disadvantage** relative to interfaces.
+
+#### Interfaces
+
+Compile-time assertions can be made in provider code that a struct instance satisfies a particular interface. For example, the following resource satisfies `ResourceFactory`:

```

Comment:
I'm not seeing `ResourceFactory` defined anywhere, and it looks like it's not being asserted against, which I think is what you're talking about here? Having trouble connecting this example to the description.

Oh, you mean that it will give a compile-time error if `computeInstanceResource` doesn't fulfill `framework.Resource`? Should we call that out explicitly? A lot of code going on there.

Reply:
Thanks, I'd accidentally mixed two examples!